### PR TITLE
[ACCEPTANCE] Scenario: Create a folder inside another folder

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
Implementing first acceptance scenario
Scenario: Create a folder inside another folder

## Related Issue
- refs #682 

## Motivation and Context
Quite simple: we need tests to protect the project from regressions and ensure to meet the acceptance criteria

## How Has This Been Tested?
- run the acceptance tests local
- or watch :robot: 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

File: tests/acceptance/stepDefinitions/loginContext.js [hunk 1]
```diff
@@ -1,13 +1,18 @@
 const { client } = require('nightwatch-api')
-const { Given, Then, When } = require('cucumber')
-// respect custom credentials via env vars
-const OC_USER = process.env.OC_USER || null
```
Comment (tempelgogo) [inline tests/acceptance/stepDefinitions/loginContext.js]: Why did you remove this ? I'm unable to use testing against my instance because my password isn't admin...
Don't want to use docker for every single acceptance test run :unamused:

File: nightwatch.conf.js [hunk 2]
```diff
@@ -1,22 +1,35 @@
...
+      globals: {
+        waitForConditionTimeout: 10000,
+        waitForConditionPollInterval: 10000
       }
     },
     local: {
       launch_url: localLaunchUrl,
+      globals: {
+        backend_url: localBackendUrl,
+        backend_admin_username: 'admin',
```
Comment (tempelgogo) [inline nightwatch.conf.js]: no override possible ? :(

File: tests/acceptance/stepDefinitions/loginContext.js (hunk 1 above)
Comment (DeepDiver1975) [reply tests/acceptance/stepDefinitions/loginContext.js]: will bring it back then .... maybe moving this to the nightwatch config is much more reasonable .... just like the lauchurl

File: nightwatch.conf.js [hunk 3]
```diff
@@ -1,23 +1,25 @@
...
+LOCAL_BACKEND_URL = LOCAL_BACKEND_URL.startsWith('http') ? LOCAL_BACKEND_URL : 'http://' + LOCAL_BACKEND_URL
 
 module.exports = {
   page_objects_path: './tests/acceptance/pageObjects',
   test_settings: {
     default: {
       globals: {
-        waitForConditionTimeout: 10000,
-        waitForConditionPollInterval: 10000
+        waitForConditionTimeout: 5000,
```
Comment (DeepDiver1975) [inline nightwatch.conf.js]: we need the higher values in drone afaik

File: tests/acceptance/features/2-webUIFiles/createFolders.feature [hunk 4]
```diff
@@ -3,11 +3,6 @@ Feature: create folders
   I want to create folders
   So that I can organise my data structure
 
-  Background:
```
Comment (DeepDiver1975) [inline tests/acceptance/features/2-webUIFiles/createFolders.feature]: reason for this?

File: tests/acceptance/stepDefinitions/loginContext.js [hunk 5]
```diff
@@ -29,8 +29,8 @@ Then('the user logs in with username {string} and password {string} using the we
     const loginPage = client.page.ownCloudLoginPage()
     return loginPage
       .waitForElementVisible('@usernameInput')
-      .setValue('@usernameInput', username)
-      .setValue('@passwordInput', password)
+      .setValue('@usernameInput', client.globals.backend_admin_username)
```
Comment (DeepDiver1975) [inline tests/acceptance/stepDefinitions/loginContext.js]: This is not correct - username and password are parameters of this step

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 6]
```diff
@@ -1,10 +1,65 @@
...
+    .click('@newFileMenuButton')
+    .waitForElementVisible('@newFolderButton')
+    .click('@newFolderButton')
+    .waitForElementVisible('@newFolderInput')
+    .setValue('@newFolderInput', folderName)
+    .click('@newFolderOkButton')
+
+  return client
+    .useXpath()
+    .waitForElementVisible('//div[contains(@class, "file-row-name")][text()="'+folderName+'"]')
```
Comment (paurakhsharma) [inline tests/acceptance/stepDefinitions/filesContext.js]: I think we can put this XPath in pageObject aswell.

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 7]
```diff
@@ -1,10 +1,65 @@
...
+    .useCss()
+});
+
+When('the user opens folder {string} using the webUI', function (folder) {
+  return client
+    .url(client.launchUrl + '/#/files/list/' + folder)
+    .waitForElementVisible('#files-list-progress')
+    .waitForElementNotPresent('#files-list-progress')
+    .waitForElementVisible('#breadcrumb-0')
+    .assert.containsText("#breadcrumb-0", folder)
```
Comment (paurakhsharma) [inline tests/acceptance/stepDefinitions/filesContext.js]: Please keep these ids in pageObject.

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 8]
```diff
@@ -1,10 +1,65 @@
...
+Then(/there should be no files\/folders listed on the webUI/, function () {
+  return client.elements('css selector','#files-list .file-row', function (result) {
+    client.assert.equal(result.value.length, 0);
+  });
+});
+
+Then('folder {string} should be listed on the webUI', function (folder) {
+  return client
+    .useXpath()
+    .waitForElementVisible('//div[contains(@class, "file-row-name")][text()="'+folder+'"]')
```
Comment (paurakhsharma) [inline tests/acceptance/stepDefinitions/filesContext.js]: Please use pageObject

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 9]
```diff
@@ -1,10 +1,65 @@
...
+    .useCss()
+});
+
+When('the user reloads the current page of the webUI', function () {
+  return client.refresh()
```
Comment (paurakhsharma) [inline tests/acceptance/stepDefinitions/filesContext.js]: We should think about creating a `waitTillPageIsLoaded()` function for each pageObject so, that we wait till the page is completely loaded after we call this step.
Something like https://github.com/owncloud/core/blob/master/tests/acceptance/features/bootstrap/WebUIGeneralContext.php#L532

File: tests/acceptance/stepDefinitions/filesContext.js (hunk 6 above)
Comment (paurakhsharma) [reply tests/acceptance/stepDefinitions/filesContext.js]: Also, this step doesn't work if the folder is not created for some reason.(e.g invalid folder name). So, we cannot use this `When` step in such scenarios as it is the job of `Then` step to verify the folder is created.

**Comment by DeepDiver1975** [comment]
THX a lot @paurakhsharma - I totally agree with all your comments on the page object. I'll address this once I know why this damn test is not properly running on drone ....

**Comment by individual-it** [comment]
@paurakhsharma moved some logic and the selectors to pageObjects, please have a look at it
I think all logic that 'attached' to the page should be in the pageObject. e.g. opening a folder, uploading a file, deleting, renaming etc.
The pageObject should know how to do that, so if the UI changes the pageObject has to be adjusted but the context and feature files stay the same
